src/extra/components/adq.py

```diff
@@ -0,0 +1,1302 @@
+
+        if not fractional:
+            pulse_period = int(round(pulse_period, 1))
+
+        return self._clock_ratio * pulse_period
+
+    def correct_common_mode(self, data, cm_period, baseline, baselevel):
+        """Correct common mode.
+
+        For ADQ digitizers, a common mode is present for every Nth
+        sample due to the parallel readout architecture, typically with
+        N = 8 or N = 16 with interleaving. This method allows to correct
+        this behaviour by computing a baseline for every Nth sample up
+        to the configured period within the baseline region and substract
+        it from every Nth pixel across the entire trace. This will change
+        the native data type to float32 if none is specified explicitly
+        or an output array passed. Note that this type of common mode
+        correction will often not yield useful results for integer types.
+
+        Args:
+            data (ArrayLike): Input data to preprocess, will be
+                converted to np.ndarray currently.
+            out (ArrayLike, optional): Result array, a new is allocated
+                if omitted.
+            cm_period (int, optional): Apply common mode correction with
+                specified sample periodicity if greater than 0, by
+                default 8 or 16 when interleaved.
+            baseline (slice or ArrayLike, optional): Contiguous 1D slice
+                of the trace of each train to determine common mode or
+                baseline or direct baseline data to use, :1000 by default.
+            baselevel (number, optional): ADU value to pull the baseline
+                to, None by default. Note that common mode corrections
+                if enabled always pull the baselevel to zero unless
+                specified otherwise here.
+
+        Returns:
+            out (ndarray): Preprocessed input data, converted to ndarray
+                if a different array was passed.
+        """
+
+        if cm_period < 1:
+            raise ValueError('Common mode must be positive number')
+
+        out = np.zeros_like(data, dtype=np.float32)
+
+        self._correct_cm_by_train(data, out, cm_period, baseline, baselevel)
+
+        return out
+
+    def pull_to_baselevel(self, data, baseline, baselevel):
+        out = np.zeros_like(data, dtype=np.float32)
+
+        self._pull_to_baselevel(data, out, baseline, baselevel)
+
+        return out
+
+    def reshape_to_pulses(self, data, pulse_ids, first_pulse_offset=None):
+        """Reshape train data to pulse data.
+
+        This method performs pulse separation by splitting the trace
+        acquired by train into individual traces by pulse based on the
+        pulse information the component is initialized with.
+
+        Args:
+            data (ArrayLike): Digitizer trace(s) for one or more trains,
+                last axis is assumed to be sample axis and on a train
+                boundary.
```

Comment:
Not sure what 'on a train boundary' means here.

Reply:
Data by train (as opposed to by pulse), but maybe that wording is clearer then.

Reply:
OK, got it. What about "last axis is assumed to be samples within a train"

Reply:
Yes, that is probably easier to understand than my attempt at "concatenated pulse traces".